AU2/plugins/custom_plugins/PolicePlugin.py

```diff
@@ -88,122 +93,205 @@ def __init__(self):
-                        identifier=self.html_ids["Rank"],
-                        title="Promote police? Must be a number from 0-9",
-                        default=e.pluginState.get(self.identifier, {}).get("rank", {}),
-                    )
-                ]
-            )
-        ]
+    def on_event_request_update(self, e: Event) -> List[HTMLComponent]:
+        # TODO Make a selector with police filtering
+        if self.gsdb_get("Manual Rank"):
+            return [
+                Dependency(
+                    dependentOn="CorePlugin_assassin_pseudonym",
+                    htmlComponents=[
+                        AssassinDependentIntegerEntry(
+                            pseudonym_list_identifier="CorePlugin_assassin_pseudonym",
+                            identifier=self.html_ids["Relative Rank"],
+                            title="Select Police to adjust rank (relative to current rank)",
+                            default=e.pluginState.get(self.identifier, {})
+                        )
+                    ]
+                )
+            ]
+        else:
+            return []
 
     def on_event_update(self, e: Event, htmlResponse) -> List[HTMLComponent]:
-        e.pluginState.setdefault(self.identifier, {})["rank"] = htmlResponse[self.html_ids["Rank"]]
-        return [Label("[POLICE] Success!")]
+        if self.gsdb_get("Manual Rank"):
+            for player_id, relative_rank in htmlResponse[self.html_ids["Relative Rank"]].items():
+                e.pluginState.setdefault(self.identifier, {})[player_id] = relative_rank
+            return [Label("[POLICE] Success!")]
+        else:
+            return [Label("[POLICE] Didn't need to do anything")]
 
-    def on_page_generate(self, htmlResponse) -> List[HTMLComponent]:
-        pass
-        
+    def on_page_generate(self, _) -> List[HTMLComponent]:
         events = list(EVENTS_DATABASE.events.values())
         events.sort(key=lambda event: event.datetime)
 
-        police_rank_manager = PoliceRankManager()
+        police_rank_manager = PoliceRankManager(auto_ranking=self.gsdb_get("Auto Rank"), police_kill_ranking=self.gsdb_get("Police Kills Rankup"))
         death_manager = DeathManager(perma_death=True)
         for e in events:
             police_rank_manager.add_event(e)
             death_manager.add_event(e)
 
+        # Code to generate new ranks if police are promoted/demoted more than ever before
+        if police_rank_manager.get_min_rank() < -int(self.gsdb_get("Default Rank")):
+            current_ranks = self.gsdb_get("Ranks")
+            current_default = int(self.gsdb_get("Default Rank"))
+            for i in range(-current_default - police_rank_manager.get_min_rank()):
+                current_ranks.insert(0, f"Level {-(i+current_default+1)} Constable")
+            self.gsdb_set("Ranks", current_ranks)
+            self.gsdb_set("Default Rank", -police_rank_manager.get_min_rank())
```

Comment:
I don't think you need to commit more ranks to the database. How about using a function like this to read from `ranks` to avoid having to mess with the DB?

```python
def getRank(idx: int, ranks: List[str]) -> str:
    if idx >= len(ranks) - 2:
        return ranks[len(ranks) - 3]
    elif idx < 0:
        return ranks[0]
    return ranks[idx]
```

Reply:
The idea behind adding them to the DB was that it would be obvious which new ranks need to be renamed when the umpire configures ranks again, while still ranking up police automatically if the umpire forgets to create the ranks in advance

Reply:
I see, that makes sense... could you please give a warning to the umpire (as a LABEL etc.) about this and leave a comment in code?